interface plugin: GTP-U tunnel support added

mtirpak opened the pull request:
Initial implementation of GTP-U tunnel interface configuration added.
Currently VPP 20.01-324 and 20.01-379 are supported only.

mtirpak committed:
interface plugin: GTP-U tunnel support added

Initial implementation of GTP-U tunnel interface configuration added.
Currently VPP 20.01-324 and 20.01-379 are supported only.

mejedi commented on plugins/vpp/ifplugin/descriptor/interface.go:
```diff
@@ -508,6 +522,13 @@ func (d *InterfaceDescriptor) Validate(key string, intf *interfaces.Interface) e
 
 		}
+	case interfaces.Interface_GTPU_TUNNEL:
+		if intf.GetGtpu().SrcAddr == "" {
```
VXLAN branch above parses IP addresses and returns `InvalidValueError` if parse fails. But we merely compare with an empty string. Shouldn't we validate if the string is a valid IP address?

mtirpak replied:
Actually only the VXLAN implementation does this. The address is checked later in the vpp interface implementation, which is a more detailed check: the address classes must match and the addresses should not be equal.
If you prefer, I can add a check also here.

mejedi commented on plugins/vpp/ifplugin/vppcalls/vpp1908/gtpu_vppcalls.go:
```diff
@@ -0,0 +1,19 @@
+
+// AddGtpuTunnel adds new GTPU interface.
+func (h *InterfaceVppHandler) AddGtpuTunnel(ifName string, gtpuLink *ifs.GtpuLink, vrf uint32, multicastIf uint32) (uint32, error) {
+    err := errors.New("Not implemented")
```
Are we capable of creating GTPU tunnels w/o implementing this function?

mejedi commented on plugins/vpp/ifplugin/vppcalls/vpp2001_379/dump_interface_vppcalls.go:
```diff
@@ -921,6 +927,58 @@ func (h *InterfaceVppHandler) dumpGreDetails(interfaces map[uint32]*vppcalls.Int
+		}
+
+		if gtpuDetails.IsIPv6 == 1 {
+			interfaces[gtpuDetails.SwIfIndex].Interface.Link = &ifs.Interface_Gtpu{
```
I'm somewhat uncomfortable with the amount of duplicated code in then-else branches.

Does it make sense to introduce a helper function for binary to text IP address conversion, like `IPString(gtpuDetails, gtpuDetails.SrcAddress)`?

mtirpak replied:
OK, let me refactor the if-else a bit.
Regarding the conversion function, this is what the net.IP module already does. Calling another function would not be too much shorted in my opinion.

mejedi commented on plugins/vpp/ifplugin/vppcalls/vpp2001_379/dump_interface_vppcalls.go:
```diff
@@ -921,6 +927,58 @@ func (h *InterfaceVppHandler) dumpGreDetails(interfaces map[uint32]*vppcalls.Int
+					Multicast:  multicastIfName,
+					SrcAddr:    net.IP(gtpuDetails.SrcAddress).To16().String(),
+					DstAddr:    net.IP(gtpuDetails.DstAddress).To16().String(),
+				    EncapVrfId: gtpuDetails.EncapVrfID,
```
Indentation looks weird. Is it how they do it in Go?

mtirpak replied:
I will still have to format the code.

mejedi commented on plugins/vpp/ifplugin/vppcalls/vpp2001_379/gtpu_vppcalls.go:
```diff
@@ -0,0 +1,79 @@
+    } else if srcAddr.To16() != nil && dstAddr.To16() != nil {
+        req.IsIPv6 = 1
+        req.SrcAddress = []byte(srcAddr.To16())
+        req.DstAddress = []byte(dstAddr.To16())
```
Please fix indentation.

mejedi commented on plugins/vpp/ifplugin/vppcalls/vpp2001_379/gtpu_vppcalls.go:
```diff
@@ -0,0 +1,79 @@
+		return 0, err
+	}
+
+	if gtpuLink.SrcAddr == gtpuLink.DstAddr {
```
Why is the source address equal to dest address considered an error?

mtirpak replied:
A tunnel with the same src and dst address does not make sense. Then there is no need to tunnel.

mejedi commented on plugins/vpp/ifplugin/descriptor/interface_crud.go:
```diff
@@ -153,6 +153,26 @@ func (d *InterfaceDescriptor) Create(key string, intf *interfaces.Interface) (me
+			if !found {
+				err = errors.Errorf("failed to find multicast interface %s referenced by GTPU %s",
+					multicastIf, intf.Name)
+				d.log.Error(err)
```
I doubt if we should log the error. The caller receives the error and she could do whatever works best for her, say logging.

mtirpak replied:
I was following the convention of the existing implementation. The error is logged in all the other cases as well.

mejedi commented on plugins/vpp/ifplugin/vppcalls/vpp2001_324/dump_interface_vppcalls.go:
```diff
@@ -921,6 +927,58 @@ func (h *InterfaceVppHandler) dumpGreDetails(interfaces map[uint32]*vppcalls.Int
 	return nil
 }
 
+// dumpGtpuDetails dumps GTP-U interface details from VPP and fills them into the provided interface map.
```
It looks like `plugins/vpp/ifplugin/vppcalls/vpp2001_324` and `plugins/vpp/ifplugin/vppcalls/vpp2001_379` are almost identical with minor alteration. For these reasons, we had to implement GTPU changes twice.

Is there any better way? What is Ligato authors' take on this?

We could avoid code duplication by only supporting the latest (379) version.

mtirpak replied:
I do not like this either, the code is duplicated indeed everywhere, I do not think we can do too much. I guess support for vpp2001_X will be dropped at some point when vpp 20.01 becomes stable and there will be only a vpp2001 subfolder.

mtirpak committed:
gtpu: Code review comments addressed

- GTPU tunnel src/dst addresses are validated before writing into
the kvs

- if-else in dump interface call is restructured.

mtirpak committed:
gtpu: encap vrf id fixed

The encapsulated (outer) VRF id can be explicitly set to a different
value than the interface VRF id. This allows the tunneled traffic
to be routed by a separate FIB table.
This is necessary because the VPP gtpu plugin implements a check
to verify whether the ingress interface vrf assignment matches
the encap vrf id of the tunnel. If not, the packet is dropped.
Without a different VRF value, it would not be possible to implement
an overlay network with separate routing tables.

Dependencies for the GTPU VRF id and multicast interface are added.

The VRF id is inherited from the unnumbered interface only if
the value is not explicitly set to non-0. This allows the tunnel
interface to re-use the IP address of another interface from a different
VRF table.

mtirpak committed:
gtpu: source code formatted

mtirpak committed:
licence note added

mtirpak committed:
gtpu: vppcalls test cases added

IP address conversion fixed: an IPv4 address was successfully
parsed as IPv6 by mistake.

mtirpak committed:
gtpu: added support for vpp 19.08

mtirpak committed:
gtpu: added support for vpp 19.04

mtirpak committed:
gtpu: added test cases for dump interface function

mtirpak committed:
gtpu: added integration tests

These tests depend on the VPP bugfix bd0a00a45637c1dde533ef6c3798418c2ab15009

mtirpak committed:
gtpu: revert the change with the VRF overwrite

This change could affect other interfaces as well, better to revert,
even if VPP supports this case.

The scenario can be made work with configuring a loopback interface
with the desired VRF id and using that interface as the tunnel
unnumbered int.

mtirpak committed:
gtpu: disable interface dump in the broken VPP versions

Disable the dump gtpu interface API call in the gtpu integration
test with some of the VPP versions that do not include the fix yet.

mtirpak committed:
gtpu: golint fix, renamed dumpApiOk to dumpAPIOk
